transport/grpc/handler.go

```diff
@@ -61,19 +60,121 @@ func (h *handler) handle(srv interface{}, serverStream grpc.ServerStream) error
 		return errInvalidGRPCStream
 	}
 
+	streamMethod := stream.Method()
+	transportRequest, err := h.getBasicTransportRequest(ctx, streamMethod)
+	if err != nil {
+		return err
+	}
+
+	handlerSpec, err := h.i.router.Choose(ctx, transportRequest)
+	if err != nil {
+		return err
+	}
+	switch handlerSpec.Type() {
+	case transport.Unary:
+		return h.handleUnary(ctx, transportRequest, serverStream, streamMethod, start, handlerSpec.Unary())
+	case transport.Streaming:
+		return toGRPCStreamError(h.handleStream(ctx, transportRequest, serverStream, start, handlerSpec.Stream()))
+	}
+	return yarpcerrors.Newf(yarpcerrors.CodeUnimplemented, "transport grpc does not handle %s handlers", handlerSpec.Type().String())
+}
+
+// getBasicTransportRequest converts the grpc request metadata into a
+// transport.Request without a body field.
+func (h *handler) getBasicTransportRequest(ctx context.Context, streamMethod string) (*transport.Request, error) {
+	md, ok := metadata.FromIncomingContext(ctx)
+	if md == nil || !ok {
+		return nil, yarpcerrors.Newf(yarpcerrors.CodeInternal, "cannot get metadata from ctx: %v", ctx)
+	}
+	transportRequest, err := metadataToTransportRequest(md)
+	if err != nil {
+		return nil, err
+	}
+
+	procedure, err := procedureFromStreamMethod(streamMethod)
+	if err != nil {
+		return nil, err
+	}
+
+	transportRequest.Procedure = procedure
+	if err := transport.ValidateRequest(transportRequest); err != nil {
+		return nil, err
+	}
+	return transportRequest, nil
+}
+
+// procedureFromStreamMethod converts a GRPC stream method into a yarpc
+// procedure name.  This is mostly copied from the GRPC-go server processing
+// logic here:
+// https://github.com/grpc/grpc-go/blob/d6723916d2e73e8824d22a1ba5c52f8e6255e6f8/server.go#L931-L956
+func procedureFromStreamMethod(streamMethod string) (string, error) {
+	if streamMethod != "" && streamMethod[0] == '/' {
+		streamMethod = streamMethod[1:]
+	}
+	pos := strings.LastIndex(streamMethod, "/")
+	if pos == -1 {
+		return "", errInvalidGRPCMethod
+	}
+	service := streamMethod[:pos]
+	method := streamMethod[pos+1:]
+	return procedureToName(service, method)
+}
+
+func (h *handler) handleStream(
+	ctx context.Context,
+	transportRequest *transport.Request,
+	serverStream grpc.ServerStream,
+	start time.Time,
+	streamHandler transport.StreamHandler,
+) error {
+	tracer := h.i.t.options.tracer
+	if tracer == nil {
+		tracer = opentracing.GlobalTracer()
+	}
+	var parentSpanCtx opentracing.SpanContext
+	md, ok := metadata.FromIncomingContext(ctx)
+	if ok {
+		parentSpanCtx, _ = tracer.Extract(opentracing.HTTPHeaders, mdReadWriter(md))
+	}
+	extractOpenTracingSpan := &transport.ExtractOpenTracingSpan{
```

Comment:
I don't know if we want to add this right now but the HTTP inbound also adds
the following tags: "rpc.caller", "rpc.service", "rpc.encoding",
"rpc.transport".

Reply:
will add in a follow-up